File: Assets/Imports/EckyECS/Components/ComponentAllocator.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using UnityEngine;
using UnityEngine.Assertions;
using UnityEngine.Profiling;

public class ComponentAllocator
{
    public static Dictionary<Type, int> TypeToID = new();
    public static Dictionary<int, Type> IDToType = new();
    public static Dictionary<Type, int> TypeToSize = new();


    public static int GetIDFor(Type Type)
    {
        Register(Type);
        return TypeToID[Type];
    }

    public static Type GetTypeFor(int ID)
    {
        Assert.IsTrue(IDToType.ContainsKey(ID));
        return IDToType[ID];
    }

    private static void Register(Type Type)
    {
        if (TypeToID.ContainsKey(Type))
            return;

        TypeToID.Add(Type, CURRENT_COMP_INDEX++);
        IDToType.Add(TypeToID[Type], Type);
        TypeToSize.Add(Type, GetSize(Type));
    }

    public static int GetSize(Type Type)
    {
        return Marshal.SizeOf(Type);
    }

    private static int CURRENT_COMP_INDEX = 0;
}

File: ImportsLib/src/EckyCS/Util/ComponentAllocator.cpp
#include "ComponentAllocator.h"
#include "../Components/Component.h"

namespace EckyCS
{
    int ComponentAllocator::CURRENT_COMP_INDEX = 0;
    std::map<std::string, int> ComponentAllocator::TypeToID;
    std::map<int, std::string> ComponentAllocator::IDToType;

    string ComponentAllocator::Register(const shared_ptr<Component>& Component)
    {
        string Type = typeid(Component.get()).name();
        if (TypeToID.contains(Type))
            return Type;

        return Register(Type);
    }

    string ComponentAllocator::Register(const std::string& Type)
    {
        if (TypeToID.contains(Type))
            return Type;
        
        TypeToID[Type] = ++CURRENT_COMP_INDEX;
        IDToType[TypeToID[Type]] = Type;
        return Type;
    }

    int ComponentAllocator::GetIDFor(const shared_ptr<Component>& Component)
    {
        auto&& Type = Register(Component);
        return TypeToID[Type];
    }

    int ComponentAllocator::GetIDFor(const std::string& Name)
    {
        if (!TypeToID.contains(Name))
            return -1;
        
        return TypeToID[Name];
    }

    string ComponentAllocator::GetTypeFor(int ID)
    {
        if (!IDToType.contains(ID))
            return "";
        
        return IDToType[ID];
    }

    
}
